src/controller/mod.rs

```diff
@@ -1482,124 +1354,545 @@ impl Drop for ControllerInner {
-    pub fn install_recipe(&self, new_recipe: String) {
-        self.rpc("install_recipe", &new_recipe).unwrap()
+    pub fn install_recipe(&mut self, new_recipe: String) {
+        self.rpc("install_recipe", &new_recipe)
     }
 
     /// graphviz description of the dataflow graph
-    pub fn graphviz(&self) -> String {
-        self.rpc("graphviz", &()).unwrap()
+    pub fn graphviz(&mut self) -> String {
+        self.rpc("graphviz", &())
     }
 
-    /// Set the `Logger` to use for internal log messages.
-    pub fn log_with(&mut self, _: slog::Logger) {}
+    /// Wait for associated local controller to exit.
+    pub fn wait(mut self) {
+        self.local.take().unwrap().1.join().unwrap()
+    }
+}
+impl<A: Authority> Drop for ControllerHandle<A> {
+    fn drop(&mut self) {
+        if let Some((sender, join_handle)) = self.local.take() {
+            let _ = sender.send(ControlEvent::Shutdown);
+            join_handle.join().unwrap();
+        }
+    }
+}
+
+/// Wrapper around the state of a thread serving network requests. Both tracks the address that the
+/// thread is listening on and provides a way to stop it.
+struct ServingThread {
+    addr: SocketAddr,
+    join_handle: JoinHandle<()>,
+    stop: Box<Drop + Send>,
+}
+impl ServingThread {
+    fn stop(self) {
+        drop(self.stop);
+        self.join_handle.join().unwrap();
+    }
+}
+
+/// Describes a running controller instance. A serialized version of this struct is stored in
+/// ZooKeeper so that clients can reach the currently active controller.
+#[derive(Serialize, Deserialize)]
+struct ControllerDescriptor {
+    external_addr: SocketAddr,
+    internal_addr: SocketAddr,
+    checktable_addr: SocketAddr,
+    nonce: u64,
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+struct ControllerConfig {
+    sharding: Option<usize>,
+    partial_enabled: bool,
+    domain_config: DomainConfig,
+    persistence: PersistenceParameters,
+    heartbeat_every: Duration,
+    healthcheck_every: Duration,
+    local_workers: usize,
+    nworkers: usize,
+}
+impl Default for ControllerConfig {
+    fn default() -> Self {
+        Self {
+            #[cfg(test)]
+            sharding: Some(2),
+            #[cfg(not(test))]
+            sharding: None,
+            partial_enabled: true,
+            domain_config: DomainConfig {
+                concurrent_replays: 512,
+                replay_batch_timeout: time::Duration::from_millis(1),
+                replay_batch_size: 32,
+            },
+            persistence: Default::default(),
+            heartbeat_every: Duration::from_secs(1),
+            healthcheck_every: Duration::from_secs(10),
+            #[cfg(test)]
+            local_workers: 2,
+            #[cfg(not(test))]
+            local_workers: 0,
+            nworkers: 0,
+        }
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+struct ControllerState {
+    config: ControllerConfig,
+    epoch: Epoch,
+    recipe: (), // TODO: store all relevant state here.
```

Comment:
This TODO doesn't affect the ability to run, because `ControllerInner` still stores the recipe in its `recipe` field, correct? AIUI, leaving this blank just means we don't persist the recipe across controller failures yet.

Reply:
Exactly.